File: dotnet/Nitro/Geography/Projections/MercatorProjection.cs
using Nitro.Geometry;

namespace Nitro.Geography.Projections
{
    public class MercatorProjection : Projection
    {
        public DoubleVector2 Project(Coordinate coordinate)
        {
            double latRad = coordinate.Latitude * Math.PI / 180.0;

            DoubleVector2 result = new DoubleVector2();
            result.X = (coordinate.Longitude + 180.0) / 360.0;
            result.Y = (1 - Math.Log(Math.Tan(latRad) + (1.0 / Math.Cos(latRad))) / Math.PI) / 2;
            
            return result;
        }

        public Coordinate Unproject(DoubleVector2 point)
        {
            double xNom = Math.Exp(2 * Math.PI) - Math.Exp(4 * Math.PI * point.Y);
            double xDen = Math.Exp(2 * Math.PI) + Math.Exp(4 * Math.PI * point.Y);

            double yNom = 2 * Math.Exp(-Math.PI * (-1 + 2 * point.Y));
            double yDen = Math.Exp(-2 * Math.PI * (-1 + 2 * point.Y)) + 1;

            Coordinate result = new Coordinate();
            result.Latitude = Math.Atan2(xNom / xDen, yNom / yDen) * 180.0 / Math.PI;
            result.Longitude = -180.0 + (360.0 * point.X);
            
            return result;
        }

        /// <summary>
        /// Calculates projected coordinates for the given zoom level.
        /// Plays well with Geography.Mapping.Descriptor
        /// </summary>
        /// <param name="coordinate">Coordinate to project</param>
        /// <param name="zoom">Target zoom level</param>
        /// <param name="x">x of the projected coordinate</param>
        /// <param name="y">y of the projected coordinate</param>
        public DoubleVector2 ToDescriptorCoordinate(Coordinate coordinate, int zoom)
        {
            DoubleVector2 result = Project(coordinate);

            double multiplyer = Math.Pow(2, zoom);

            result.X *= multiplyer;
            result.Y *= multiplyer;

            return result;
        }

        /// <summary>
        /// Calculates a WGS84 coordinate from a projected one. The reverse function of
        /// ToDescriptorCoordinate.
        /// </summary>
        /// <param name="x"></param>
        /// <param name="y"></param>
        /// <param name="zoom"></param>
        /// <returns></returns>
        public Coordinate FromDescriptorCoordinate(double x, double y, int zoom)
        {
            double multiplyer = Math.Pow(2, zoom);

            x /= multiplyer;
            y /= multiplyer;

            return Unproject(new DoubleVector2(x, y));
        }
    }
}

File: web/webmap/src/geography/MercatorProjection.ts
import { DoubleVector2 } from "../geometry/DoubleVector2";
import { Coordinate } from "./Coordinate";
import { Projection } from "./Projection";

export class MercatorProjection implements Projection {
    public project(coordinate: Coordinate): DoubleVector2 {
        const latRad: number = (coordinate.latitude * Math.PI) / 180.0;

        const result: DoubleVector2 = new DoubleVector2();
        result.x = (coordinate.longitude + 180.0) / 360.0;
        result.y = (1 - Math.log(Math.tan(latRad) + 1.0 / Math.cos(latRad)) / Math.PI) / 2;

        return result;
    }

    public unproject(point: DoubleVector2): Coordinate {
        const xNom: number = Math.exp(2 * Math.PI) - Math.exp(4 * Math.PI * point.y);
        const xDen: number = Math.exp(2 * Math.PI) + Math.exp(4 * Math.PI * point.y);

        const yNom: number = 2 * Math.exp(-Math.PI * (-1 + 2 * point.y));
        const yDen: number = Math.exp(-2 * Math.PI * (-1 + 2 * point.y)) + 1;

        const result: Coordinate = new Coordinate();
        result.latitude = (Math.atan2(xNom / xDen, yNom / yDen) * 180.0) / Math.PI;
        result.longitude = -180.0 + 360.0 * point.x;

        return result;
    }

    public toDescriptorCoordinate(coordinate: Coordinate, zoom: number): DoubleVector2 {
        const result: DoubleVector2 = this.project(coordinate);

        const multiplier: number = Math.pow(2, zoom);

        result.x *= multiplier;
        result.y *= multiplier;

        return result;
    }

    public fromDescriptorCoordinate(x: number, y: number, zoom: number): Coordinate {
        const multiplier: number = Math.pow(2, zoom);

        x /= multiplier;
        y /= multiplier;

        return this.unproject(new DoubleVector2(x, y));
    }
}